lib/methwhy.g

```diff
@@ -158,108 +244,172 @@ local oper,narg,args,skip,verbos,fams,flags,i,j,methods,flag,flag2,
+        nopargs," arguments:\n");
+  Print("#I  Total: ",nummeth," entries, of which ",numapplic); 
+  if numapplic=1 then 
+    Print(" is applicable:\n");
+  else 
+    Print(" are applicable:\n");
+  fi;
+  if nr>numapplic then 
+    if (numapplic=0) then 
+      Print("#I  there are no valid methods with these parameters\n"); 
+    else
+      if numapplic=1 then 
+        Print("#I  there is only 1 valid method\n"); 
+      else
+        Print("#I  there are only ",numapplic," valid methods\n"); 
+      fi;
+    fi;
   fi;
-  if verbos > 0 then 
-    Print("#I  Total: ", Length(methods)," entries\n");
+  ## find the range [first..last] of methods which might be printed
+  if nr=0 then  ## this is the case "all" 
+    first:=1;
+    last:=nummeth;
+  elif numapplic=0 then 
+    if verbos=1 then 
+      return fail; 
+    else
+      first:=1;
+      last:=nummeth;
+    fi;
+  elif nr>numapplic then 
+    first:=1;
+    last:=0;
+  else 
+    last:=applic[nr]; 
+    if verbos>1 then 
+      if nr=1 then 
+        first:=1;
+      else
+        first:=applic[nr-1]+1; 
+      fi;
+    else 
+      first := applic[nr];
+    fi;
   fi;
-  for i in [1..Length(methods)] do
+  ## loop over the methods to be printed
+  for i in [first..last] do
     m := methods[i];
     nam:=m.info;
     val:=m.rank;
     oper:=m.func;
-    if verbos>1 then
-      Print("#I  Method ",i,": ``",nam,"''");
+    if i in applic then 
+      Print("#I  Method ",i); 
+      Print(", valid operation number ",Position(applic,i)); 
+      Print(", value: ");  
+      Print_Value_SFF(val);
+      Print("\n#I  ``",nam,"''\n");
       if IsBound(m.location) then
-        Print(" at ", m.location[1], ":", m.location[2]);
+        Print("#I   at ", m.location[1], ":", m.location[2]);
       elif LocationFunc(oper) <> "" then
         Print(" at ",LocationFunc(oper));
       fi;
-      Print(", value: ");
-      Print_Value_SFF(val);
       Print("\n");
-    fi;
-    flag:=true;
-    j:=1;
-    while j<=narg and (flag or verbos>3) do
-      if j=1 and isconstructor then
-	flag2:=IS_SUBSET_FLAGS(m.argFilt[j],flags[j]);
-      else
-	flag2:=IS_SUBSET_FLAGS(flags[j],m.argFilt[j]);
-      fi;
-      flag:=flag and flag2;
-      if flag2=false and verbos>2 then
-	need:=NamesFilter(m.argFilt[j]);
-	if j=1 and isconstructor then
-	  Print("#I   - ",Ordinal(j)," argument must be ",
-		need,"\n");
-	else
-	  has:=NamesFilter(flags[j]);
-	  Print("#I   - ",Ordinal(j)," argument needs ",
-		Filtered(need,i->not i in has),"\n");
-	fi;
+    elif verbos>1 then
+      Print("#I  Method ",i); 
+      Print(", value: ");  
+      Print_Value_SFF(val);
+      Print("\n#I  ``",nam,"''\n");
+      if IsBound(m.location) then
+        Print("#I   at ", m.location[1], ":", m.location[2]);
+      elif LocationFunc(oper) <> "" then
+        Print(" at ",LocationFunc(oper));
       fi;
-      j:=j+1;
-    od;
-    if flag then
-      if fams=fail or CallFuncList(m.famPred,fams) then
```

Comment:
it looks as if the test for the family predicate has been removed and not been replaced properly.

Reply:
This test appears to be needed only for methods which are applicable.  Do you know of an example which triggers the Print("#I   - bad family relations\n"); warning?